File: python/leetcode/string/trie/1233_remove_subfolder.py
"""
1233. Remove Sub-Folders from the Filesystem (Medium)

Given a list of folders, remove all sub-folders in those folders and return in any order the folders after removing.

If a folder[i] is located within another folder[j], it is called a sub-folder of it.

The format of a path is one or more concatenated strings of the form: / followed by one or more lowercase English letters. For example, /leetcode and /leetcode/problems are valid paths while an empty string and / are not.

Example 1:

Input: folder = ["/a","/a/b","/c/d","/c/d/e","/c/f"]
Output: ["/a","/c/d","/c/f"]
Explanation: Folders "/a/b/" is a subfolder of "/a" and "/c/d/e" is inside of folder "/c/d" in our filesystem.
Example 2:

Input: folder = ["/a","/a/b/c","/a/b/d"]
Output: ["/a"]
Explanation: Folders "/a/b/c" and "/a/b/d/" will be removed because they are subfolders of "/a".
Example 3:

Input: folder = ["/a/b/c","/a/b/ca","/a/b/d"]
Output: ["/a/b/c","/a/b/ca","/a/b/d"]
 

Constraints:

1 <= folder.length <= 4 * 10^4
2 <= folder[i].length <= 100
folder[i] contains only lowercase letters and '/'
folder[i] always starts with character '/'
Each folder name is unique.
"""

class TrieNode():
	def __init__(self):
		self.end = False
		self.next = {}

class Solution(object):
    def removeSubfolders(self, folder):
        """
        :type folder: List[str]
        :rtype: List[str]
        """
        folder.sort(key=lambda x:len(x))
        root = TrieNode()
        result = []
        for f in folder:
        	tmplist = f.split("/")
        	# print(len(tmplist), tmplist)
        	node = root
        	flag = True
        	for i, tmp in enumerate(tmplist):
        		if node.end:
        			flag = False
        			break
        		if tmp not in node.next:
        		    node.next[tmp] = TrieNode()
        		node = node.next[tmp]
        		if i == len(tmplist) - 1:
        			node.end = True
        	if flag:
        		result.append(f)
        return result



if __name__ == "__main__":
	a = Solution()
	print(a.removeSubfolders(["/a","/a/b","/c/d","/c/d/e","/c/f"]))
	print(a.removeSubfolders(["/a","/a/b/c","/a/b/d"]))
	print(a.removeSubfolders(["/a/b/c","/a/b/ca","/a/b/d"]))

File: cc/leetcode/string/trie/1233_remove_subfolder.cc

class Node {
public:
    Node() {end=false;}
    unordered_map<string, Node*> memo;
    bool end;
};

class Solution {
public:
    struct less_key {
        bool operator() (const string& s1, const string& s2) {
            return s1.size() < s2.size();
        }
    };
    
    void parse(string& s, vector<string>& dir) {
        int i = 0, j;
        while (i < s.size()) {
            if (s[i] == '/')
                i++;
            j = i;
            while (j < s.size() && s[j] != '/')
                j++;
            dir.push_back(s.substr(i, j-i));
            i = j + 1;
        }
    }
    
    vector<string> removeSubfolders(vector<string>& folder) {
        // sort by length
        sort(folder.begin(), folder.end(), less_key());
        
        vector<string> result;
        Node* root = new Node();
        Node* n;
        for (string& s : folder) {
            vector<string> tmp;
            parse(s, tmp);
            /*cout << s <<":";
            for (int i = 0; i < tmp.size(); ++i)
                cout << tmp[i] << " ";
            cout << endl;*/
            bool flag = true;
            n = root;
            for (int i = 0; i < tmp.size(); ++i) {
                if (n->end) {
                    flag = false;
                    break;
                }
                string item = tmp[i];
                if (n->memo.find(item) == n->memo.end())
                    n->memo[item] = new Node();
                n = n->memo[item];
                if (i == tmp.size() - 1)
                    n->end = true;
            }
            if (flag)
                result.push_back(s);
        }
        
        return result;
    }
};
